### PR TITLE
Update 'skip_dir' and 'skip_file' pattern matching and handling (Issue #982)

### PR DESCRIPTION
**Pull request by abraunegg**
* Update 'skip_dir' and 'skip_file' pattern matching and handling

### CODE REVIEW

**Comment by abraunegg** [comment]
**client configuration:**
```
Configuration file successfully loaded
onedrive version                       = v2.4.3-5-gab5599b
Config path                            = /home/alex/.config/onedrive-personal/
Config file found in config path       = true
Config option 'check_nosync'           = false
Config option 'sync_dir'               = /home/alex/OneDrivePersonal
Config option 'skip_dir'               = /Images|/Music|.dropbox.cache
Config option 'skip_file'              = ~*|.~*|*.tmp|*.swp|SMPP*.pdf|/shared/Cisco VDI Whitepaper/white_paper_c11-494994.pdf
Config option 'skip_dotfiles'          = false
Config option 'skip_symlinks'          = false
Config option 'monitor_interval'       = 20
Config option 'min_notify_changes'     = 5
Config option 'log_dir'                = /var/log/onedrive/
Config option 'classify_as_big_delete' = 1000
Config option 'sync_root_files'        = false
Selective sync 'sync_list' configured  = false
Business Shared Folders configured     = false
```
**Expectation:**
* Any file matching *.tmp will be skipped
* Any file matching SMPP*.pdf will be skipped
* Explicit file /shared/Cisco VDI Whitepaper/white_paper_c11-494994.pdf will be skipped
* Folders with explicit path of /Images , /Music will skipped
* Any folder called .dropbox.cache in any path will be skipped

**Current Master:**
```
./onedrive --confdir '~/.config/onedrive-personal/' --synchronize --verbose --resync
Using 'user' Config Dir: /home/alex/.config/onedrive-personal/
Using 'system' Config Dir: 
Configuration file successfully loaded
Deleting the saved status ...
Initializing the OneDrive API ...
Configuring Global Azure AD Endpoints
Opening the item database ...
All operations will be performed in: /home/alex/OneDrivePersonal
Application version: v2.4.3-5-gab5599b
Account Type: personal
Default Drive ID: 66d53be8a5056eca
Default Root ID: 66D53BE8A5056ECA!101
Remaining Free Space: 5368288511
Fetching details for OneDrive Root
OneDrive Root does not exist in the database. We need to add it.
Added OneDrive Root to the local database
Initializing the Synchronization Engine ...
Syncing changes from OneDrive ...
Applying changes of Path ID: 66D53BE8A5056ECA!101
Processing 22 OneDrive items to ensure consistent local state
Creating local directory: level 1
Creating local directory: shared
Downloading file Images/valgrind-out-202005170812.txt ... done.
Creating local directory: level 1/level 2
Creating local directory: Images/child
Creating local directory: Images/Child 34
Creating local directory: level 1/level 2/level 3
Downloading file Documents/Images/qwerqasdfasdf.jpg ... done.
Downloading file Documents/Music/ewrqqwerqwer.mp3 ... done.
Creating local directory: Images/child/asdfasdf
Downloading file Images/child/asdfasdf/valgrind-out-202005170812.txt ... done.
Processing 1 OneDrive items to ensure consistent local state due to a full scan being triggered by actions on OneDrive
Syncing this OneDrive Personal Shared Folder: shared
Applying changes of Path ID: BC7D88EC1F539DCF!1288
Processing 7 OneDrive items to ensure consistent local state
Creating local directory: shared/Cisco VDI Whitepaper
Creating local directory: shared/SMPP
Downloading file shared/Cisco VDI Whitepaper/white_paper_c11-494994.pdf ... done.
Downloading file shared/Cisco VDI Whitepaper/campover.pdf ... done.
Skipping item - excluded by skip_file config: SMPP-IF-SPEC_v3_3-24858.pdf
Skipping item - excluded by skip_file config: SMPP_v3_4_Issue1_2-24857.pdf
Uploading differences of ~/OneDrivePersonal
Processing root
The directory has not changed
Processing random_files
The directory has not changed
Processing random_images
The directory has not changed
Processing random_videos
The directory has not changed
Processing level 1
The directory has not changed
Processing level 2
The directory has not changed
Processing level 3
The directory has not changed
Processing Documents
The directory has not changed
Processing Images
The directory has not changed
Processing qwerqasdfasdf.jpg
The file has not changed
Processing Music
The directory has not changed
Processing ewrqqwerqwer.mp3
The file has not changed
Processing Images
The directory has not changed
Processing valgrind-out-202005170812.txt
The file has not changed
Processing child
The directory has not changed
Processing asdfasdf
The directory has not changed
Processing valgrind-out-202005170812.txt
The file has not changed
Processing Child 34
The directory has not changed
Processing .Images
The directory has not changed
Processing Dropbox
The directory has not changed
Processing shared
The directory has not changed
Processing shared
The directory has not changed
Processing Cisco VDI Whitepaper
The directory has not changed
Processing white_paper_c11-494994.pdf
The file has not changed
Processing campover.pdf
The file has not changed
Processing SMPP
The directory has not changed
Uploading new items of ~/OneDrivePersonal
Skipping item - excluded by skip_dir config: ./Dropbox/.dropbox.cache
Uploading new file ./Documents/Images/image001.jpg ... done.
Remaining free space on OneDrive: 5368288486
Uploading new file ./Documents/Music/audio001.mp3 ... done.
Remaining free space on OneDrive: 5368288461
OneDrive Client requested to create remote path: ./Documents/.Images
The requested directory to create was not found on OneDrive - creating remote directory: ./Documents/.Images
Successfully created the remote directory ./Documents/.Images on OneDrive
OneDrive Client requested to create remote path: ./Documents/.Music
The requested directory to create was not found on OneDrive - creating remote directory: ./Documents/.Music
Successfully created the remote directory ./Documents/.Music on OneDrive
Uploading new file ./Images/image002.jpg ... done.
Remaining free space on OneDrive: 5368288436
OneDrive Client requested to create remote path: ./Music
The requested directory to create was not found on OneDrive - creating remote directory: ./Music
Successfully created the remote directory ./Music on OneDrive
Uploading new file ./Music/audio002.mp3 ... done.
Remaining free space on OneDrive: 5368288411
OneDrive Client requested to create remote path: ./.Music
The requested directory to create was not found on OneDrive - creating remote directory: ./.Music
Successfully created the remote directory ./.Music on OneDrive
Skipping item - excluded by skip_file config: ./file1.tmp
Applying changes of Path ID: 66D53BE8A5056ECA!101
Processing 30 OneDrive items to ensure consistent local state due to a full scan being triggered by actions on OneDrive
Processing 1 OneDrive items to ensure consistent local state due to a full scan being requested
Syncing this OneDrive Personal Shared Folder: shared
Applying changes of Path ID: BC7D88EC1F539DCF!1288
```

**With Patch:**
```
./onedrive --confdir '~/.config/onedrive-personal/' --synchronize --verbose --resync 
Using 'user' Config Dir: /home/alex/.config/onedrive-personal/
Using 'system' Config Dir: 
Configuration file successfully loaded
Deleting the saved status ...
Initializing the OneDrive API ...
Configuring Global Azure AD Endpoints
Opening the item database ...
All operations will be performed in: /home/alex/OneDrivePersonal
Application version: v2.4.3-5-gab5599b
Account Type: personal
Default Drive ID: 66d53be8a5056eca
Default Root ID: 66D53BE8A5056ECA!101
Remaining Free Space: 5368288411
Fetching details for OneDrive Root
OneDrive Root does not exist in the database. We need to add it.
Added OneDrive Root to the local database
Initializing the Synchronization Engine ...
Syncing changes from OneDrive ...
Applying changes of Path ID: 66D53BE8A5056ECA!101
Processing 30 OneDrive items to ensure consistent local state
Creating local directory: ./level 1
Creating local directory: ./shared
Skipping item - excluded by skip_dir config match: /Images
Skipping item - excluded by skip_dir config match: /Music
Creating local directory: ./level 1/level 2
Creating local directory: ./level 1/level 2/level 3
Downloading file Documents/Images/qwerqasdfasdf.jpg ... done.
Downloading file Documents/Music/ewrqqwerqwer.mp3 ... done.
The local item has a different modified time 2020-Jul-06 03:10:22Z remote is 2020-Jul-06 03:08:06Z
The local item has a different modified time 2020-Jul-06 03:10:22Z remote is 2020-Jul-06 03:08:06Z
Processing 1 OneDrive items to ensure consistent local state due to a full scan being triggered by actions on OneDrive
Syncing this OneDrive Personal Shared Folder: shared
Applying changes of Path ID: BC7D88EC1F539DCF!1288
Processing 7 OneDrive items to ensure consistent local state
Creating local directory: ./shared/Cisco VDI Whitepaper
Creating local directory: ./shared/SMPP
Skipping item - excluded by skip_file config: white_paper_c11-494994.pdf
Downloading file shared/Cisco VDI Whitepaper/campover.pdf ... done.
Skipping item - excluded by skip_file config: SMPP-IF-SPEC_v3_3-24858.pdf
Skipping item - excluded by skip_file config: SMPP_v3_4_Issue1_2-24857.pdf
Uploading differences of ~/OneDrivePersonal
Processing root
The directory has not changed
Processing random_files
The directory has not changed
Processing random_images
The directory has not changed
Processing random_videos
The directory has not changed
Processing level 1
The directory has not changed
Processing level 2
The directory has not changed
Processing level 3
The directory has not changed
Processing .Images
The directory has not changed
Processing Dropbox
The directory has not changed
Processing shared
The directory has not changed
Processing shared
The directory has not changed
Processing Cisco VDI Whitepaper
The directory has not changed
Processing campover.pdf
The file has not changed
Processing SMPP
The directory has not changed
Processing Documents
The directory has not changed
Processing .Images
The directory has not changed
Processing .Music
The directory has not changed
Processing Images
The directory has not changed
Processing qwerqasdfasdf.jpg
The file has not changed
Processing image001.jpg
The file last modified time has changed
Processing Music
The directory has not changed
Processing ewrqqwerqwer.mp3
The file has not changed
Processing audio001.mp3
The file last modified time has changed
Processing .Music
The directory has not changed
Uploading new items of ~/OneDrivePersonal
Skipping item - excluded by skip_dir config: ./Dropbox/.dropbox.cache
Skipping item - excluded by skip_dir config: ./Images
Skipping item - excluded by skip_dir config: ./Music
Skipping item - excluded by skip_file config: ./file1.tmp
Applying changes of Path ID: 66D53BE8A5056ECA!101
Processing 30 OneDrive items to ensure consistent local state due to a full scan being triggered by actions on OneDrive
Skipping item - excluded by skip_dir config match: /Images
Skipping item - excluded by skip_dir config match: /Music
Processing 1 OneDrive items to ensure consistent local state due to a full scan being requested
Syncing this OneDrive Personal Shared Folder: shared
Applying changes of Path ID: BC7D88EC1F539DCF!1288
```

**Comment by abraunegg** [comment]
@norbusan 
Whilst this issue was 'supposed' to be 'fixed' by https://github.com/abraunegg/onedrive/commit/dcc0322cac451ec2297f7df03e770a978d94e268 - it actually broke the full path pattern matching.

This patch, based on my testing, is more complete and does not break the pattern matching when trying to 'skip_dir' dot folders which was the issue with #901.

File: src/sync.d [hunk 1]
```diff
@@ -2171,7 +2236,7 @@ final class SyncEngine
 			break;
 		case ItemType.dir:
 		case ItemType.remote:
-			log.log("Creating local directory: ", path);
+			log.log("Creating local directory: ", "./" ~ path);
```
Comment (norbusan) [inline src/sync.d:2239]: I'm a bit surprised about this one here. Is this necessary? Do we chdir into the respective directories?

Comment (abraunegg) [reply src/sync.d:2239]: With this one - when creating local directories, the following was being printed:
```
Creating local directory: level 1/level 2
Creating local directory: Images/child
Creating local directory: Images/Child 34
```
With this PR, that now becomes:
```
Creating local directory: ./level 1
Creating local directory: ./shared
Skipping item - excluded by skip_dir config match: /Images
Skipping item - excluded by skip_dir config match: /Music
Creating local directory: ./level 1/level 2
Creating local directory: ./level 1/level 2/level 3
```
It was more so output consistency .. can remove this as really non essential - will do so then merge the PR.

**Pull request merged**